pkg/virt-operator/resource/apply/apps.go

```diff
@@ -87,7 +109,140 @@ func (r *Reconciler) syncDeployment(origDeployment *appsv1.Deployment) (*appsv1.
+	}
+	return newDs, nil
+}
+
+func (r *Reconciler) getCanaryPod(daemonSet *appsv1.DaemonSet) *corev1.Pod {
+	for _, obj := range r.stores.InfrastructurePodCache.List() {
+		pod := obj.(*corev1.Pod)
+		owner := metav1.GetControllerOf(pod)
+
+		if owner != nil && owner.Name == daemonSet.Name && util.PodIsUpToDate(pod, r.kv) {
+			return pod
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) howManyUpdatedAndReadyPods(daemonSet *appsv1.DaemonSet) int32 {
+	var updatedReadyPods int32
+
+	for _, obj := range r.stores.InfrastructurePodCache.List() {
+		pod := obj.(*corev1.Pod)
+		owner := metav1.GetControllerOf(pod)
+
+		if owner != nil && owner.Name == daemonSet.Name && util.PodIsUpToDate(pod, r.kv) && util.PodIsReady(pod) {
+			updatedReadyPods++
+		}
+	}
+	return updatedReadyPods
+}
+
+func daemonHasDefaultRolloutStrategy(daemonSet *appsv1.DaemonSet) bool {
+	return getMaxUnavailable(daemonSet) == daemonSetDefaultMaxUnavailable.IntValue()
+}
+
+func (r *Reconciler) processCanaryUpgrade(cachedDaemonSet, newDS *appsv1.DaemonSet, forceUpdate bool) (bool, error, CanaryUpgradeStatus) {
+	var updatedAndReadyPods int32
+
+	isDaemonSetUpdated := util.DaemonSetIsUpToDate(r.kv, cachedDaemonSet) && !forceUpdate
+	desiredReadyPods := cachedDaemonSet.Status.DesiredNumberScheduled
+	if isDaemonSetUpdated {
+		updatedAndReadyPods = r.howManyUpdatedAndReadyPods(cachedDaemonSet)
+	}
+
+	switch {
+	case isDaemonSetUpdated && updatedAndReadyPods == 0:
+		// check for a crashed canary pod
+		canary := r.getCanaryPod(cachedDaemonSet)
+		if canary != nil && util.PodIsCrashLooping(canary) {
+			r.recorder.Eventf(cachedDaemonSet, corev1.EventTypeWarning, failedUpdateDaemonSetReason, "daemonSet %v rollout failed", cachedDaemonSet.Name)
+			return false, fmt.Errorf("daemonSet %s rollout failed", cachedDaemonSet.Name), CanaryUpgradeStatusFailed
+		}
+		return false, nil, CanaryUpgradeStatusStarted
+	case updatedAndReadyPods == 0:
+		// start canary upgrade
+		setMaxUnavailable(newDS, intstr.FromInt(1))
+		_, err := r.patchDaemonSet(cachedDaemonSet, newDS)
+		if err != nil {
+			return false, fmt.Errorf("unable to start canary upgrade for daemonset %+v: %v", newDS, err), CanaryUpgradeStatusFailed
+		}
+		return false, nil, CanaryUpgradeStatusStarted
+	case updatedAndReadyPods < desiredReadyPods && daemonHasDefaultRolloutStrategy(cachedDaemonSet):
+		// canary was ok, start real rollout
+		setMaxUnavailable(newDS, daemonSetFastMaxUnavailable)
+		// start rollout again
+		_, err := r.patchDaemonSet(cachedDaemonSet, newDS)
+		if err != nil {
+			return false, fmt.Errorf("unable to update daemonset %+v: %v", newDS, err), CanaryUpgradeStatusFailed
+		}
+		log.Log.V(2).Infof("daemonSet %v updated", newDS.GetName())
+		return false, nil, CanaryUpgradeStatusUpgradingDaemonSet
+	case updatedAndReadyPods < desiredReadyPods:
+		log.Log.V(4).Infof("waiting for all pods of daemonSet %v to be ready", newDS.GetName())
+		return false, nil, CanaryUpgradeStatusWaitingDaemonSetRollout
+	}
```

Comment:
I'm having trouble with this switch statement. Particularly around the cases where no action is taking place.

If we're going to use a switch statement, can the cases map directly to the steps? Something like... 

```
case updatedAndReadyPods == 0:
   // set max unavailable to 1 if it isn't already
   // else no-op and wait 
  return false
case updatedAndReadyPods > 0 && updatedAndReadyPods < desiredReadyPods 
  // set max unavailable to 10% if it isn't already
  // else no-op and wait
  return false
case updatedAndReadyPods > 0 && updatedAndReadyPods == desiredReadyPods
  // set max unavailable to 1 if it isn't already
  // set generation
  // daemonset is fully online, return so return true
  return true
```

Reply:
Updated, does this look better?